lib/spanner/config/upgrader.ex

```diff
@@ -0,0 +1,130 @@
+      :ok ->
+        do_upgrade(config)
+        |> insert_deprecation_msg(deprecation_msg)
+      {:error, errors} ->
+        {:error, errors, [deprecation_msg]}
+    end
+  end
+  # If we get an un-upgradable version, let the user know.
+  def upgrade(%{"cog_bundle_version" => version}) do
+    {:error,
+     [{"""
+       cog_bundle_version #{version} is not supported. \
+       Please update your bundle config to version #{@current_version}.\
+       """,
+       "#/cog_bundle_version"}],
+     []}
+  end
+  # If we don't get a version, let the user know.
+  def upgrade(_) do
+    {:error,
+     [{"""
+       cog_bundle_version not specified. You must specify a valid bundle \
+       version. The current version is #{@current_version}.\
+       """,
+       "#/cog_bundle_version"}],
+     []}
+  end
+
+  defp do_upgrade(config) do
+    case execution_once?(config["commands"]) do
+      {true, {warnings, errors}} ->
+        {:error, errors, warnings}
+      {false, warnings} ->
+        {enforce_warnings, updated_commands} = update_enforcing(config["commands"])
+        updated_config = %{config | "commands" => updated_commands}
+        |> Map.put("cog_bundle_version", @current_version)
+        {:ok, updated_config, warnings ++ enforce_warnings}
+    end
+  end
+
+  # Checks to see if the execution field exists in the bundle config. If it
+  # does but contains "multiple" we just return a warning. If it contains
+  # "once" we return an error.
+  defp execution_once?(commands) do
+    {warnings, errors} = Enum.reduce(commands, {[],[]},
+      fn
+        ({cmd_name, cmd}, {warnings, errors}) ->
+          case Map.get(cmd, "execution", nil) do
+            "once" ->
+              msg = "Execution 'once' commands are no longer supported"
+              location = "#/commands/#{cmd_name}/execution"
+              updated_errors = [{msg, location} | errors]
+              {warnings, updated_errors}
+            "multiple" ->
+              msg = "Execution type has been deprecated. Please update your bundle config"
+              location = "#/commands/#{cmd_name}/execution"
+              updated_warnings = [{msg, location} | warnings]
+              {updated_warnings, errors}
+            nil ->
+              {warnings, errors}
+          end
+      end)
+
+    if length(errors) > 0 do
+      {true, {warnings, errors}}
+    else
+      {false, warnings}
+    end
+
+  end
+
+  # Updates for non enforcing commands. If 'enforcing: false' is specified we
+  # add the "allow" rule, delete the enforcing field and return a warning. If
+  # 'enforcing: true' is specified we return a warning and delete the field.
+  defp update_enforcing(commands) do
+    Enum.reduce(commands, {[], commands},
+      fn
+        ({cmd_name, %{"enforcing" => enforcing}=cmd}, {warnings, commands}) when not(enforcing) ->
+          msg = "Non-enforcing commands have been deprecated. Please update your bundle config"
```

Comment:
Bundle config version errors should explicitly mention which bundle config version to use.